File: src/my_project/counterpoint/skeleton_generator.py
import random
from dataclasses import dataclass

from my_project.model import Duration, Key, Melody, Note, NoteName, PartId, Pitch


@dataclass(frozen=True)
class Skeleton:
    """
    課題全体を仮に全音符で実施した際の、協和音の音高と和音の設定を表す。
    """

    measures: list[Melody[Note[Pitch, list[NoteName]]]]
    # TODO: attribute は Chord[Degree] なのかで迷っている


@dataclass
class SkeletonGenerator:
    cantus_firmus: list[Pitch]
    key: Key
    part_id: PartId
    rand: random.Random
    _measure_length: int

    # TODO Iterator でさまざまなスケルトンを返すようにする
    def generate_skeleton(self) -> Skeleton:
        """
        CFを元に実施する声部のそれぞれの小節の最初に利用する和声音を定める。
        この和声音を骨格として、間に旋律を埋めることで課題が実施される。
        そのため、ここでは旋律の同音の連続や7度などの不協和音程を選択することができる。

        和音と旋律の規則。和音はCFと協和する
        - 冒頭小節
            - 和音は I の基本形
            - 対位旋律は i, v 度音から始める。(移勢の類(?)の場合は iii 度音が可能)
            - CFと1度になることが可能
        - 途中の小節
            - 長調の場合、 I, II, III, IV, V, VI の和音の基本形および TODO
        - 最終小節の1小節前
            - V の基本形、第一転回形
            - VII の第一転回形
            - II の基本形、第一転回形(!)。ただし対位旋律は経過音として導音を経過する場合に限られる。
        - 最終小節
            - 和音は I の基本形
            - 対位旋律は i 度音 (2声の場合)
            - CFと1度になることが可能
        """

        # TODO: ひとまずダミーのものを一つ返すようにしている
        if self.key == Key.parse("C Major") and self.cantus_firmus == [
            Pitch.parse("C4"),
            Pitch.parse("A3"),
            Pitch.parse("G3"),
            Pitch.parse("C3"),
        ]:
            skeleton = Skeleton(
                measures=[
                    _create_measure(
                        Pitch.parse("G4"),
                        [NoteName.parse(n) for n in ["C", "E", "G"]],
                    ),
                    _create_measure(
                        Pitch.parse("A4"),
                        [NoteName.parse(n) for n in ["A", "C", "E"]],
                    ),
                    _create_measure(
                        Pitch.parse("B4"),
                        [NoteName.parse(n) for n in ["G", "B", "D"]],
                    ),
                    _create_measure(
                        Pitch.parse("C5"),
                        [NoteName.parse(n) for n in ["C", "E", "G"]],
                    ),
                ]
            )
            return skeleton
        elif self.key == Key.parse("C Major") and self.cantus_firmus == [
            Pitch.parse("C4"),
            Pitch.parse("A3"),
            Pitch.parse("G3"),
            Pitch.parse("E3"),
            Pitch.parse("F3"),
            Pitch.parse("A3"),
            Pitch.parse("G3"),
            Pitch.parse("E3"),
            Pitch.parse("D3"),
            Pitch.parse("C3"),
        ]:
            skeleton = Skeleton(
                measures=[
                    _create_measure(
                        Pitch.parse("G4"),
                        [NoteName.parse(n) for n in ["C", "E", "G"]],
                    ),
                    _create_measure(
                        Pitch.parse("A4"),
                        [NoteName.parse(n) for n in ["A", "C", "E"]],
                    ),
                    _create_measure(
                        Pitch.parse("B4"),
                        [NoteName.parse(n) for n in ["G", "B", "D"]],
                    ),
                    _create_measure(
                        Pitch.parse("C5"),
                        [NoteName.parse(n) for n in ["C", "E", "G"]],
                    ),
                    _create_measure(
                        Pitch.parse("D5"),
                        [NoteName.parse(n) for n in ["D", "F", "A"]],
                    ),
                    _create_measure(
                        Pitch.parse("F5"),
                        [NoteName.parse(n) for n in ["F", "A", "C"]],
                    ),
                    _create_measure(
                        Pitch.parse("G5"),
                        [NoteName.parse(n) for n in ["G", "B", "D"]],
                    ),
                    _create_measure(
                        Pitch.parse("C5"),
                        [NoteName.parse(n) for n in ["C", "E", "G"]],
                    ),
                    _create_measure(
                        Pitch.parse("B4"),
                        [NoteName.parse(n) for n in ["D", "F", "B"]],
                    ),
                    _create_measure(
                        Pitch.parse("C5"),
                        [NoteName.parse(n) for n in ["C", "E", "G"]],
                    ),
                ]
            )
            return skeleton
        elif self.key == Key.parse("A Minor") and self.cantus_firmus == [
            Pitch.parse("A2"),
            Pitch.parse("B2"),
            Pitch.parse("C3"),
            Pitch.parse("E3"),
            Pitch.parse("F3"),
            Pitch.parse("E3"),
            Pitch.parse("C3"),
            Pitch.parse("A2"),
            Pitch.parse("B2"),
            Pitch.parse("A2"),
        ]:
            skeleton = Skeleton(
                measures=[
                    _create_measure(
                        Pitch.parse("E3"),
                        [NoteName.parse(n) for n in ["A", "C", "E"]],
                    ),
                    _create_measure(
                        Pitch.parse("G#3"),
                        [NoteName.parse(n) for n in ["B", "D", "G#"]],
                    ),
                    _create_measure(
                        Pitch.parse("A3"),
                        [NoteName.parse(n) for n in ["A", "C", "E"]],
                    ),
                    _create_measure(
                        Pitch.parse("G#3"),
                        [NoteName.parse(n) for n in ["E", "G#", "B"]],
                    ),
                    _create_measure(
                        Pitch.parse("D4"),
                        [NoteName.parse(n) for n in ["D", "F", "A"]],
                    ),
                    _create_measure(
                        Pitch.parse("E4"),
                        [NoteName.parse(n) for n in ["E", "G#", "B"]],
                    ),
                    _create_measure(
                        Pitch.parse("A4"),
                        [NoteName.parse(n) for n in ["A", "C", "E"]],
                    ),
                    _create_measure(
                        Pitch.parse("C4"),
                        [NoteName.parse(n) for n in ["A", "C", "E"]],  # or F A C
                    ),
                    _create_measure(
                        Pitch.parse("G#3"),
                        [NoteName.parse(n) for n in ["B", "D", "G#"]],
                    ),
                    _create_measure(
                        Pitch.parse("A3"),
                        [NoteName.parse(n) for n in ["A", "C", "E"]],
                    ),
                ]
            )
            return skeleton
        elif self.key == Key.parse("D Minor") and self.cantus_firmus == [
            Pitch.parse("D3"),
            Pitch.parse("E3"),
            Pitch.parse("F3"),
            Pitch.parse("A3"),
            Pitch.parse("Bb3"),
            Pitch.parse("A3"),
            Pitch.parse("F3"),
            Pitch.parse("D3"),
            Pitch.parse("E3"),
            Pitch.parse("D3"),
        ]:
            skeleton = Skeleton(
                measures=[
                    _create_measure(
                        Pitch.parse("A3"),
                        [NoteName.parse(n) for n in ["D", "F", "A"]],
                    ),
                    _create_measure(
                        Pitch.parse("C#4"),
                        [NoteName.parse(n) for n in ["E", "G", "C#"]],
                    ),
                    _create_measure(
                        Pitch.parse("D4"),
                        [NoteName.parse(n) for n in ["D", "F", "A"]],
                    ),
                    _create_measure(
                        Pitch.parse("C#4"),
                        [NoteName.parse(n) for n in ["A", "C#", "E"]],
                    ),
                    _create_measure(
                        Pitch.parse("G4"),
                        [NoteName.parse(n) for n in ["G", "Bb", "D"]],
                    ),
                    _create_measure(
                        Pitch.parse("A4"),
                        [NoteName.parse(n) for n in ["A", "C#", "E"]],
                    ),
                    _create_measure(
                        Pitch.parse("D5"),
                        [NoteName.parse(n) for n in ["D", "F", "A"]],
                    ),
                    _create_measure(
                        Pitch.parse("F4"),
                        [NoteName.parse(n) for n in ["D", "F", "A"]],  # or D F Bb
                    ),
                    _create_measure(
                        Pitch.parse("C#4"),
                        [NoteName.parse(n) for n in ["E", "G", "C#"]],
                    ),
                    _create_measure(
                        Pitch.parse("D4"),
                        [NoteName.parse(n) for n in ["D", "F", "A"]],
                    ),
                ]
            )
            return skeleton
        raise NotImplementedError


def _create_measure(pitch: Pitch, chord: list[NoteName]) -> Melody[Note[Pitch, list[NoteName]]]:
    return Melody.of(Note(pitch, Duration.of(4), chord))

File: backend22/src/main/scala/composer/counterpoint/skeleton/SkeletonGenerator.scala
package composer.counterpoint.skeleton

import composer.counterpoint.model._
import composer.counterpoint.validation.Validator
import model.elements.{Degree, Duration, Key, Pitch}
import model.elements.Interval.IntervalStep
import model.elements.Degree.DegreeStep
import model.elements.Pitch.NoteName
import model.containers.{Melody, Note}
import model.elements.Key.Mode
import model.elements.Part

case class PitchWithBass(pitch: Pitch, bass: ChordWithBass[NoteName])

case class Skeleton(
    measures: List[Melody[PitchWithBass, Note[PitchWithBass]]],
)

class SkeletonGenerator(
    cantusFirmus: List[Pitch],
    key: Key,
    pitchRange: (Pitch, Pitch),
    rand: scala.util.Random,
    measureLength: Int,
) {

  private val MAX_RETRIES = 100

  def generateSkeleton(): Skeleton = {
    var attempt = 0
    while (attempt < MAX_RETRIES) {
      try {
        return generateSkeletonImpl()
      } catch {
        case e: Exception =>
          if (attempt == MAX_RETRIES - 1) throw e
      }
      attempt += 1
    }
    throw new RuntimeException("Unexpected error in generateSkeleton")
  }

  private def generateSkeletonImpl(): Skeleton = {
    var measures: List[Melody[PitchWithBass, Note[PitchWithBass]]] = Nil
    val chordPalette                                               = getChordPalette
    val candidatePitchesPool                                       = getAllPitchesInRange

    for ((cfPitch, measureIdx) <- cantusFirmus.zipWithIndex) {
      val cfDegree          = Degree.fromNoteNameKey(cfPitch.noteName, key)
      val currentMeasurePos = currentMeasurePosition(measureIdx)

      var chordCandidates: List[HarmonicDegreeChord] = Nil

      if (Set(MeasurePosition.FIRST, MeasurePosition.LAST).contains(currentMeasurePos)) {
        chordCandidates = List(DegreeChord.I)
      } else if (currentMeasurePos == MeasurePosition.PENULTIMATE) {
        if (cfDegree.step == DegreeStep.idx1(5)) {
          if (key.mode == Mode.Major) chordCandidates = List(DegreeChord.V)
          else chordCandidates = List(DegreeChord.V_leading)
        } else if (cfDegree.step == DegreeStep.idx1(2)) {
          if (key.mode == Mode.Major) chordCandidates = List(DegreeChord.II, DegreeChord.VII)
          else chordCandidates = List(DegreeChord.VII_leading)
        } else if (cfDegree.step == DegreeStep.idx1(7)) {
          if (key.mode == Mode.Major) chordCandidates = List(DegreeChord.V)
          else chordCandidates = List(DegreeChord.V_leading)
        } else {
          chordCandidates = filterChordsContaining(chordPalette, cfDegree)
        }
      } else {
        chordCandidates = filterChordsContaining(chordPalette, cfDegree)
      }

      if (chordCandidates.isEmpty) {
        chordCandidates = filterChordsContaining(chordPalette, cfDegree)
        if (chordCandidates.isEmpty) {
          throw new RuntimeException(s"No valid chord found for CF note $cfPitch ($cfDegree) at measure $measureIdx")
        }
      }

      chordCandidates = rand.shuffle(chordCandidates)
      var measureSucceeded = false

      val chordIterator = chordCandidates.iterator
      while (chordIterator.hasNext && !measureSucceeded) {
        val selectedChord  = chordIterator.next()
        val chordNoteNames = selectedChord.elements.map(_.noteName(key)).toList

        var validCandidates: List[(Pitch, ChordWithBass[NoteName])] = Nil
        for (p <- candidatePitchesPool) {
          if (chordNoteNames.contains(p.noteName)) {
            val interval          = (p - cfPitch).abs
            val intervalSemitones = interval.num.value

            if (intervalSemitones <= 24) {
              val isUnison   = intervalSemitones == 0
              var isValidPos = true

              if (currentMeasurePos == MeasurePosition.FIRST) {
                val deg = Degree.fromNoteNameKey(p.noteName, key)
                if (!Set(0, 4).contains(deg.step.value)) isValidPos = false
              } else if (currentMeasurePos == MeasurePosition.LAST) {
                val deg = Degree.fromNoteNameKey(p.noteName, key)
                if (deg.step.value != 0) isValidPos = false
              } else if (currentMeasurePos == MeasurePosition.PENULTIMATE) {
                if (isUnison) isValidPos = false
              } else {
                if (isUnison) isValidPos = false
              }

              if (isValidPos) {
                val bassPitch     = if (cfPitch.num.value < p.num.value) cfPitch else p
                val chordWithBass = ChordWithBass(HarmonicChord(chordNoteNames.toSet), bassPitch.noteName)

                val dsElements      = chordNoteNames.map(nn => Degree.fromNoteNameKey(nn, key).step).toSet
                val bassDs          = Degree.fromNoteNameKey(bassPitch.noteName, key).step
                val dsChordWithBass = DegreeStepChordWithBass(HarmonicChord(dsElements), bassDs)

                var isInversionValid = true
                try {
                  if (dsChordWithBass.inversionType == Inversion.SECOND) isInversionValid = false
                } catch {
                  case _: IllegalArgumentException => // Not a triad
                }

                if (isInversionValid) {
                  var isValidatorValid = true
                  if (measures.nonEmpty) {
                    val previousPitch   = measures.head.elems.head.value.pitch
                    val previousCfPitch = cantusFirmus(measureIdx - 1)

                    val prevMeasure = createAnnotatedMeasure(previousPitch)
                    val currMeasure = createAnnotatedMeasure(p)

                    if (!Validator.validate(Some(prevMeasure), currMeasure, Some(previousCfPitch), cfPitch)) {
                      isValidatorValid = false
                    }
                  }

                  if (isValidatorValid) {
                    validCandidates ::= (p, chordWithBass)
                  }
                }
              }
            }
          }
        }

        if (validCandidates.nonEmpty) {
          val (chosenPitch, chosenChord) = validCandidates(rand.nextInt(validCandidates.length))
          measures ::= createMeasure(chosenPitch, chosenChord)
          measureSucceeded = true
        }
      }

      if (!measureSucceeded) {
        throw new RuntimeException(
          s"No valid chord and pitch combination found for CF note $cfPitch at measure $measureIdx",
        )
      }
    }

    measures = measures.reverse

    val allAnnotatedMeasures = measures.map(m => createAnnotatedMeasure(m.elems.head.value.pitch))
    if (!Validator.validateAll(allAnnotatedMeasures)) {
      throw new RuntimeException("Failed all measure validation (e.g. total range)")
    }

    Skeleton(measures)
  }

  private def getChordPalette: List[HarmonicDegreeChord] = {
    var palette: List[HarmonicDegreeChord] = Nil

    def makeChord(roots: List[Int]): HarmonicDegreeChord = {
      DegreeChord.of(roots.map(r => getDiatonicDegree(r))*)
    }

    palette ::= DegreeChord.I
    palette ::= DegreeChord.II
    palette ::= makeChord(List(3, 5, 7))
    palette ::= makeChord(List(4, 6, 8)) // 8 is 1 octave up

    if (key.mode == Mode.Major) palette ::= DegreeChord.V
    else palette ::= DegreeChord.V_leading

    palette ::= makeChord(List(6, 8, 10))

    if (key.mode == Mode.Major) palette ::= DegreeChord.VII
    else palette ::= DegreeChord.VII_leading

    palette.reverse
  }

  private def getDiatonicDegree(stepIdx1: Int): Degree = {
    val intervalStep = IntervalStep.idx_1(stepIdx1)
    val p            = key.diatonicScalePitch(intervalStep)
    Degree.fromNoteNameKey(p.noteName, key)
  }

  private def filterChordsContaining(pool: List[HarmonicDegreeChord], degree: Degree): List[HarmonicDegreeChord] = {
    pool.filter(_.elements.contains(degree))
  }

  private def getAllPitchesInRange: List[Pitch] = {
    val (minP, maxP)         = pitchRange
    var pitches: List[Pitch] = Nil
    for (octVal <- (minP.octave.value - 1) to (maxP.octave.value + 1)) {
      for (nnVal <- -15 to 19) {
        val p = Pitch.of(octVal, nnVal)
        if (p.num.value >= minP.num.value && p.num.value <= maxP.num.value) {
          pitches ::= p
        }
      }
    }
    pitches.reverse
  }

  private def currentMeasurePosition(currentMeasureIdx: Int): MeasurePosition = {
    if (currentMeasureIdx == 0) MeasurePosition.FIRST
    else if (currentMeasureIdx == measureLength - 1) MeasurePosition.LAST
    else if (currentMeasureIdx == measureLength - 2) MeasurePosition.PENULTIMATE
    else MeasurePosition.MIDDLE
  }

  private def createMeasure(
      pitch: Pitch,
      chord: ChordWithBass[NoteName],
  ): Melody[PitchWithBass, Note[PitchWithBass]] = {
    val note = Note(PitchWithBass(pitch, chord), Duration.of(4), Part.Root)
    Melody(List(note))
  }

  private def createAnnotatedMeasure(pitch: Pitch): AnnotatedMeasure = {
    val note = Note(
      AnnotatedNote(Some(pitch), NoteAnnotation(isTiedStart = false, toneType = ToneType.HARMONIC_TONE)),
      Duration.of(4),
      Part.Root,
    )
    Melody(List(note))
  }
}
